tfjs-backend-webgpu/src/depthwise_conv2d_nchw_shared_webgpu.ts

```diff
@@ -0,0 +1,177 @@
+          ${activationOp}
+        }`;
+      } else {
+        activationSnippet = `
+          fn activation(a : f32, outCoord : vec4<i32>) -> f32 {
+            ${activationOp}
+          }
+        `;
+      }
+
+      applyActivationSnippet = `dotProd = activation(dotProd, coords);`;
+    }
+
+    const addBiasSnippet = this.addBias ?
+        'dotProd = dotProd + getBiasByOutputCoords(coords);' :
+        '';
+
+    const filterSize = this.filterWidth * this.filterHeight;
+    const workGroupSize =
+        this.workGroupSize[0] * this.workGroupSize[1] * this.workGroupSize[2];
+    const tileAHeight = this.workGroupSize[1] + this.filterHeight - 1;
+    const tileAWidth = this.workGroupSize[0] + this.filterWidth - 1;
+
+    const userCode = `
+      ${activationSnippet}
+
+      var<workgroup> mm_Asub : array<array<f32, ${tileAWidth}>, ${tileAHeight}>;
+      var<workgroup> mm_Bsub : array<array<f32, ${this.filterWidth}>, ${
+        this.filterHeight}>;
+      fn readX(batch : i32, channel : i32, row : i32, col : i32) -> f32 {
+        var value = 0.0;
+        if (row >=0 && row < uniforms.inDims[0] && col >=0 && col < uniforms.inDims[1])
+        {
+          value = getX(batch, channel, row, col);
+        }
+        return value;
+      }
+
+      ${getWorkGroupSizeString()}
+      fn main(@builtin(local_invocation_id) LocalId : vec3<u32>,
+              @builtin(global_invocation_id) GlobalId : vec3<u32>,
+              @builtin(local_invocation_index) LocalIndex: u32,
+              @builtin(num_workgroups) NumWorkgroups: vec3<u32>) {
+        localId = LocalId;
+        globalId = GlobalId;
+        let localIndex = i32(LocalIndex);
+        numWorkgroups = NumWorkgroups;
+        let coords = getOutputCoords();
+        let batch = coords[0];
+        let xRCCorner = vec2<i32>(coords.zw) * uniforms.stride - uniforms.pad;
+        let d2 = coords[1];
+        let d1 = d2 / uniforms.channelMul;
+        let q = d2 - d1 * uniforms.channelMul;
+
+        let inputRowStart = xRCCorner.x;
+        let inputColStart = xRCCorner.y;
+
+        let localRow = i32(localId.y);
+        let localCol = i32(localId.x);
+
+        // Load one tile of X into local memory.
+        for (var inputRow = localRow; inputRow < ${
+        tileAHeight}; inputRow = inputRow + ${this.workGroupSize[1]}) {
+          for (var inputCol = localCol; inputCol < ${
+        tileAWidth}; inputCol = inputCol + ${this.workGroupSize[0]}) {
+            let rowOffset = inputRow - localRow;
+            let colOffset = inputCol - localCol;
+            mm_Asub[inputRow][inputCol] = readX(batch, d1, inputRowStart + rowOffset, inputColStart + colOffset);
+          }
+        }
+
+        // Load one tile of W into local memory.
+        var wIndex = localIndex;
+        ${
+        filterSize < workGroupSize ?
+            `if (wIndex < ${filterSize})` :
+            `for(; wIndex < ${filterSize}; wIndex = wIndex + ${workGroupSize})`}
```

Comment:
Is `if` statement needed here? Is it already a subset of `for`?

Reply:
I may be slightly good for perf to use `if` instead of `for` since it doesn't need to call `wIndex = wIndex + ${workGroupSize})` and another `wIndex < ${filterSize}` checking if filterSize is smaller than workGroupSize.